### ISSUE TITLE
Throw an appropriate Error (an instance of Error class), instead of Object

### ISSUE THREAD

**Issue by RomanDavlyatshin** [typescript-eslint/typescript-eslint#2400]
- [x] I have first restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```TS
// the code you're trying to parse
// any invalid TS code
hello {{{{}
```

```TS
// the code you're using to do the parse of the aforementioned code
import { parse, TSESTreeOptions } from '@typescript-eslint/typescript-estree';

export default function (source, options: TSESTreeOptions = {}, defaultOptions: TSESTreeOptions = {
  comment: false,
  jsx: true,
  loc: true,
  range: true
}) {
  return parse(source, {
    ...defaultOptions,
    ...options
  });
}

```

**Expected Result**
Error (that is an instance of Error) being thrown, with stack trace info:
```
{
   message: '';' expected',
   stack: *error stack*,
   // I suggest to extend default Error to add additional params
   details: {
     column: 6,
     index: 43,
     lineNumber: 2,
   }
}
```

**Actual Result**
I'm getting an Object instead of Error
```
{
  index: 43,
  lineNumber: 2,
  column: 6,
  message: "';' expected.",
}
```

**Additional Info**
Throwing anything besides Error is a bad practice because a stack trace is lost.
The stack trace is valuable info.
If it's crucial to add additional info I suggest extending the default Error class with the complementary field and put that info there.

At first glance, it looks like it could be easily done (create extended Error class and modify [createError](https://github.com/typescript-eslint/typescript-eslint/blob/6f397df42cbcf05c10f304c9bbfdae4803aa0ce2/packages/typescript-estree/src/node-utils.ts#L686) method to return `new CustomError({/*object containing details*/}) //error name is a subject to the discussion`)
I would like to submit a PR if no one minds, but I would be glad to hear any critique.

P.S. Of course, I could just wrap all calls to `parse` into `try ... catch` and rethrow, but I would like to see that fixed in-library

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `2.34.0` |
| `TypeScript`                           | `3.9.5` |
| `node`                                 | `14.4.0` |

**Comment by bradzacher** [comment]
This error handling can be traced all the way back to the very first commit made for this parser, back on 3 Dec 2015: https://github.com/eslint/typescript-eslint-parser/commit/c88ba98639d84ef21b10dd26ef526b5422f2d1bb#diff-6ecca2b7675f672f87297d93eec1e4baR179-R189

Nobody has ever needed a full stack trace so it's never been something anyone has changed.
5 years ago the JS landscape was vastly different, so this was a normal thing to do then.

Happy to accept a PR to change this, but note that this will come with a lot of additional testing of dependencies. It should probably be a breaking change, as there may be consumers relying on the fact that it throws this object and not an error.

cc @fisker / @sosukesuzuki - do you know if prettier has a reliance on the shape of the error?

**Comment by fisker** [comment]
It should not be a problem for Prettier.
